### PR TITLE
Version 2!

### PR DESCRIPTION
**Pull request by dehuazhu**
what is still missing (work in progress)
1) matching of dSAMu and sMu into one object
2) measure HNL reconstruction efficiency (a new Analyzer is being written)

### CODE REVIEW

File: python/physicsobjects/DisplacedMuon.py [hunk 1]
```diff
@@ -35,5 +35,7 @@ def track(self):
         ''' return reco::TrackRef '''   
         return ROOT.reco.TrackRef(self.collection_, self.position_in_collection)
 
+    def GetPhysObj(self):
```
Comment (rmanzoni) [inline python/physicsobjects/DisplacedMuon.py]: I believe this is not needed, why can't one simply do `dmuon.physObj`?

**Comment by dehuazhu** [comment]
Ur probably right...

File: python/analyzers/HNLAnalyzer.py [hunk 2]
```diff
@@ -90,75 +92,107 @@ def process(self, event):
...
-        # event.maxptdSAMu=sorted(event.dSAMu,key=lambda x: x.pt(),reverse=True)[0]
-        
-        ##############
-        # for now ONLY work with sMu
-        ##############
+       
+        # Merge Reco Muons
+        # Create an array of DisplacedMuon objects, summarizing all sMu and dSAMus into a single array, while avoiding redundancies through dR<0.2
+        dMus = []
+        dxy_cut = 1000 # cut selection for sMu / dSAMu in mm
```
Comment (rmanzoni) [inline python/analyzers/HNLAnalyzer.py]: better have it configurable. No big deal for now, just to keep the record

File: python/analyzers/HNLAnalyzer.py [hunk 3]
```diff
@@ -90,75 +92,107 @@ def process(self, event):
...
+                # if the reco particle is a displaced thing, it does not have the p4() method, so let's build it 
                 myp4 = ROOT.Math.LorentzVector('<ROOT::Math::PxPyPzE4D<double> >')(il.px(), il.py(), il.pz(), sqrt(il.mass()**2 + il.px()**2 + il.py()**2 + il.pz()**2))
                 ic = ROOT.reco.RecoChargedCandidate() # instantiate a dummy RecoChargedCandidate
                 ic.setCharge(il.charge())           # assign the correct charge
                 ic.setP4(myp4)                      # assign the correct p4
-                ic.setTrack(il.track())             # set the correct TrackRef
+                if il.reco == 'sMu':
+                    ic.setTrack(il.outerTrack())             # set the correct TrackRef
+                if il.reco == 'dSAMu':
+                    ic.setTrack(il.GetPhysObj().track())             # set the correct TrackRef
```
Comment (rmanzoni) [inline python/analyzers/HNLAnalyzer.py]: mmm, no this I think breaks the logic I had in mind.
Plus, if the muon has been reconstructed by the standard algorithm and its displacement is less than 1m, I think it's better to use the (best) track it has to offer, that is `track()` and not `outerTrack()`

**Comment by rmanzoni** [comment]
@dehuazhu I think I'll merge now and start to work on this. Ok?

File: python/analyzers/HNLAnalyzer.py (hunk 2 above)
Comment (dehuazhu) [reply python/analyzers/HNLAnalyzer.py]: What is the most "correct" way to set the tune parameters for an analysis?

**Comment by dehuazhu** [comment]
image
First "interesting" visualization of the new algorithm using a single sample:
The reconstruction efficiency of "sMu only" and "sMu & dSAMu"
Dispite the large error bars, it seems that the prior is more efficient. This might arise from a larger impurity in the latter.

**Comment by dehuazhu** [comment]
First characterization of the HNL Analyzer using full statistics (22.8 Million events):

number of total entries in chain:	22871574
number of all events with 2 or more muons:	22871574
number of events with reconstructable HNLs:	8862558
number of events with reconstructed DiMuons:	9514851
number of correctly found HNLs using Chi2 method:	2141597
number of correctly found HNLs using Dxy method:	1950699
Reconstruction efficiency (min Chi2 method):	24.2%
Reconstruction efficiency (max Dxy method):	22.0%
Reconstruction purity (min Chi2 method):	22.5%
Reconstruction purity (max Dxy method):	20.5%

The reconstruction efficiency as a function of gen_displacement is as follows:
image

with the full statics, the highest reconstruction efficiency seems to come from the max Dxy method and only using sMu... Although if considering using sMu AND dSAMu, the efficiency is higher for the min Chi2 method...

The reconstruction purity as a function of gen_displacment is as follows:
image
The Purity has the same 'ranking'

**Comment by dehuazhu** [comment]
For the current todos i've in mind, see https://github.com/dehuazhu/HNL/projects/1

**Comment by dehuazhu** [comment]
The most recent characterization of the ntuple:
screen shot 2018-07-10 at 11 21 11

**Pull request merged**